File: 22. Implement stack using two queues.py
"""
Task: Implement stack using two queues
Time Complexity: push O(n) or pop O(n) depending on approach
Space Complexity: O(n)
"""

from collections import deque

class MyStack:
    def __init__(self):
        self.q1 = deque()
        self.q2 = deque()

    def push(self, x):
        self.q1.append(x)

    def pop(self):
        while len(self.q1) > 1:
            self.q2.append(self.q1.popleft())
        val = self.q1.popleft()
        self.q1, self.q2 = self.q2, self.q1
        return val

    def top(self):
        while len(self.q1) > 1:
            self.q2.append(self.q1.popleft())
        val = self.q1.popleft()
        self.q2.append(val)
        self.q1, self.q2 = self.q2, self.q1
        return val

    def empty(self):
        return not self.q1

if __name__ == "__main__":
    s = MyStack()
    s.push(1); s.push(2); print(s.top()); print(s.pop()); print(s.empty())

# Task: Implement stack using two queues

File: 22. Implement stack using two queues.cpp
/*
 Task: Implement stack using two queues
*/

#include <bits/stdc++.h>
using namespace std;

class MyStack
{
    queue<int> q1, q2;

   public:
    void push(int x) { q1.push(x); }

    int pop()
    {
        if (q1.empty())
            throw runtime_error("empty");
        while (q1.size() > 1)
        {
            q2.push(q1.front());
            q1.pop();
        }
        int v = q1.front();
        q1.pop();
        swap(q1, q2);
        return v;
    }

    int top()
    {
        if (q1.empty())
            throw runtime_error("empty");
        while (q1.size() > 1)
        {
            q2.push(q1.front());
            q1.pop();
        }
        int v = q1.front();
        q2.push(v);
        q1.pop();
        swap(q1, q2);
        return v;
    }

    bool empty() { return q1.empty(); }
};

int main()
{
    MyStack s;
    s.push(1);
    s.push(2);
    cout << s.top() << "\n";
    cout << s.pop() << "\n";
    cout << s.empty() << "\n";
    return 0;
}

// Time Complexity: O(n) for pop/top, O(1) for push
// Space Complexity: O(n)
